### ISSUE TITLE
missing = "listwise" gives different WLS.V matrix than missing = "pairwise"

### ISSUE THREAD

**Issue by SachaEpskamp** [yrosseel/lavaan#158]
Hi Yves and others posting here!

I have been working on replicating several aspects from lavaan (http://psychonetrics.org), with the goal to estimate similar models to typical SEM models (with networks). The availability of lavaan and the access to the open source code have been fundamental for figuring some detailed things out and checking if my implementations are correct. So thanks a lot!! I also noticed how incredibly remarkable the speed and performance of lavaan is, as it is faster than some C++ implementations I made and surely faster than my R based implementations...

I'm now looking into WLS estimation with ordered categorical variables and noticed this: the WLS.V matrix reported by lavaan differs between missing = "listwise" and missing = "pairwise". See here for an example:
```r
library("lavaan")
library("mvtnorm")

# Some data:
nVar <- 10
Rho <- matrix(0.5,nVar,nVar)
diag(Rho) <- 1
Data <- as.data.frame(rmvnorm(500,sigma = Rho))
thresholds <- qnorm(c(0.25,0.5,0.75))
for (i in 1:nVar){
  Data[,i] <- as.ordered(cut(Data[,i],c(-Inf,thresholds,Inf)))
}
names(Data) <- paste0("v",1:nVar)

# Simple model:
mod <- paste(apply(combn(colnames(Data),2),2,function(x)paste0(x[1],"~~",x[2])), collapse = "\n")

# Run model:
fit_listwise <- cfa(mod, Data, missing = "listwise")
fit_pairwise <- cfa(mod, Data, missing = "pairwise")

# Extract weights matrices:
wlsv1 <- lavInspect(fit_listwise, "WLS.V")
wlsv2 <- lavInspect(fit_pairwise, "WLS.V")

# Plot:
plot(wlsv1[lower.tri(wlsv1,diag=TRUE)], wlsv2[lower.tri(wlsv2,diag=TRUE)],
     xlab = "WLS.V listwise", ylab = "WLS.V pairwise")
abline(0,1) 
```
which gives:
![image](https://user-images.githubusercontent.com/1207242/63914609-5307b600-ca66-11e9-9c66-5558e99ccac2.png)

Which means the estimates are slightly different, even though there is no missing data present. I couldn't really find a reason for this in the `lav_muthen1984.R` script. Is this intended?

On a related question, I was wondering how to properly weigh the WLS weights matrix when using pairwise observations. For full data, I now multiply the B matrix (eq 21) and the inner part of eq 22 from Muthen 1984 with 1/N, which seems to work. I tried to multiply each element of these matrices with the 1 / Neffective, with Neffective being the effective sample size for an element (say, for a threshold - correlation relation that would be the N that has no missing data on the one variable the threshold is based and the two variables the correlation is based), but this gave me some weird results not at all in line with lavaan anymore.

Best, Sacha

**Comment by yrosseel** [comment]
It turns out this is a bug. Thanks for pointing this out. Fixed now in dev 0.6-6.

This was the problem: the muthen1984() function defines a function lav_crossprod2() to handle missing values, but it (apparently) assumed 'y' was always a vector... As a result, if y was a matrix, we got the wrong values in 'A21', resulting in diagonal elements in WLS.V that are (slightly) off. This never turned up in my tests as I only used binary data. Only when multiple thresholds are needed, 'y' will be a matrix.

As for the weighting: I am not sure. If you email me an example, I will have a further look.

Yves.

**Comment by SachaEpskamp** [comment]
Hi Yves,

Thanks for this! I can now replicate the WLS matrix including with missings! 

What I meant with the weighing is that I multiply the weights matrix with 1/n:

https://github.com/SachaEpskamp/psychonetrics/blob/master/src/10_ordinal_covPrepare.cpp#L680

And I was wondering if that should not be an effective n per element. But this works instead, and I figure the constant really doesn't matter in estimation and it is cancelled out again in other functions that compute the standard errors and so.
